### ISSUE TITLE
[text-to-speech] new API doesn't work

### ISSUE THREAD

**Issue by mrwnmonm** [watson-developer-cloud/node-sdk#816]
Hi,
We are trying to migrate our text to speech implementation to the new authentication method.
We were able to obtain the access token from curl and the authentication API, then we tried to use raw websockets, but we didn't receive any data.

So we followed the example [here](https://github.com/watson-developer-cloud/node-sdk/blob/master/examples/text_to_speech_websocket.js) (which is updated 13 days ago) which still authenticate using the old method (username and password), but we received no data using that too.

Here is our code (just a dummy script):
```javascript
var TextToSpeechV1 = require('watson-developer-cloud/text-to-speech/v1');

var textToSpeech = new TextToSpeechV1({
//   iam_apikey: '<api_key>',
//   url: '<service_url>'
  username: "<username>",
  password: "<password>"
});

var synthesizeParams = {
    text: 'Hello world',
    accept: 'audio/mp3',
    voice: 'en-US_AllisonVoice'
};

async function f() {
    let promise = new Promise((resolve, reject) => {
        // Pipe the synthesized text to a file.
        var synthesizeStream = textToSpeech.synthesizeUsingWebSocket(synthesizeParams)
        
        synthesizeStream.on('message', (message, data) => {
            console.log(data)
        })
        synthesizeStream.on('close', (code, reason) => {
            console.log(code)
            resolve("100")
        })
        synthesizeStream.on('error', (err) => {
            console.log(err)
        })  
    });
  
    let result = await promise; // wait till the promise resolves (*)
    console.log(result); // "done!"
}
  
f().then(()=>{
    console.log("done")
});
```
Notice that the example uses the old authentication method, and the docs [here](https://console.bluemix.net/apidocs/text-to-speech?language=node) uses the `api_key` and the `url`, **but both methods don't work**, we can't receive data using either of them.

- Expected behavior: Receive audio data.
- Actual behavior: Receiving no data.
- Node version: 8.5.0
- SDK version: 3.15.0

**Comment by germanattanasio** [comment]
Hi @mrwnmonm, I'm sorry that you are having issues with the SDK.
I will try to take a look at this, but I'm swamped with other things so it may take me a few days.

**Comment by mrwnmonm** [comment]
Related: https://stackoverflow.com/questions/53851076/parsing-watson-tts-http-response-in-nodejs

**Comment by jeffpk62** [comment]
@germanattanasio I think we first need to determine why the example works with Cloud Foundry username/password credentials but not with a valid IAM API key.  I'm testing with SDK version 3.15.0.  My CF credentials work fine; my API key fails with the indicated error.

**Comment by mrwnmonm** [comment]
Guys, any updates on this?

**Comment by dpopp07** [comment]
@mrwnmonm @jeffpk62 I will look at this today

**Comment by dpopp07** [comment]
@mrwnmonm Your original issue is caused by two problems - one that you can fix now and one that I posted a fix for with #823.

1) Because you are not piping your stream anywhere, the underlying class needs to know to start sending data. You aren't receiving any data because the connection is not initialized. You need to add the line `synthesizeStream.resume();` after you instantiate the stream. See [this note in the example](https://github.com/watson-developer-cloud/node-sdk/blob/master/examples/text_to_speech_websocket.js#L25).

2) Once you do that, the request will be made but the connection will fail because of a bug handling the query parameters. The line `voice: 'en-US_AllisonVoice'` would cause the request to fail. Now that the fix is posted, you should be able to use as is.

@jeffpk62 Looking at the example, that is for `synthesize` over HTTP, rather than WebSockets. When the HTTP method is used without a callback, it will return a Stream for CF instances but will not return anything for IAM instances. This is a known technical limitation of handling IAM tokens in the SDK and will not be solved. Instead, we are going to start returning Promises when no callback is specified so behavior is consistent between CF and IAM.

To summarize - we should change the example to use a callback function instead of assuming that a stream is being returned.

**Comment by mrwnmonm** [comment]
@dpopp07 Thanks, it works perfectly now.  :+1: :pray: 
Of course we didn't go with piping the data, because when requesting the timings, the returned data could be the timings object or the audio buffer. So we went with `synthesizeStream.resume();`.

**Comment by DerrickYang007** [comment]
Thanks for fix the problem.But  the example code in the docs  here https://console.bluemix.net/apidocs/text-to-speech?language=node#synthesize-audio 

is wrong. It is official document example code. Please update it right.  This issure wasted me several days to find the right place and ringt  answer to  solve it. Thank you.

**Comment by germanattanasio** [comment]
Hi @DerrickYang007, we are updating the examples this week, so this will get updated. 

cc: @dpopp07, @jeffpk62

**Comment by msalimbe1** [comment]
> Thanks for fix the problem.But the example code in the docs here https://console.bluemix.net/apidocs/text-to-speech?language=node#synthesize-audio
> 
> is wrong. It is official document example code. Please update it right. This issure wasted me several days to find the right place and ringt answer to solve it. Thank you.

Same thing for me. Unfortunate.

**Comment by jeffpk62** [comment]
The examples have been updated in the source code and will be published very soon.  Sorry for the problems this caused.  One important note:  Please use the documentation at this URL:

https://cloud.ibm.com/apidocs/text-to-speech?code=node#synthesize-audio

The documentation at https://console.bluemix.net is outdated.  In the next couple of weeks, redirects will automatically take you to the current https://cloud.ibm.com documentation.

**Comment by DerrickYang007** [comment]
Thanks 




--
发自我的网易邮箱手机智能版



在 2019-04-11 19:35:25，"Jeff Kaminski" <notifications@github.com> 写道：


The examples have been updated in the source code and will be published very soon. Sorry for the problems this caused. One important note: Please use the documentation at this URL:

https://cloud.ibm.com/apidocs/text-to-speech?code=node#synthesize-audio

The documentation at https://console.bluemix.net is outdated. In the next couple of weeks, redirects will automatically take you to the current https://cloud.ibm.com documentation.

**Comment by oshihirii** [comment]
Does `synthesizeStream.resume();` only work with `synthesizeUsingWebSocket`, as commented in the link shared by @dpopp07 :

https://github.com/watson-developer-cloud/node-sdk/blob/master/examples/text_to_speech_websocket.js#L25

I am trying two approaches to get audio in a `map()` function and both giving error ` Not Found: Not Found`. 

I've come across this thread and thinking maybe it is because i am not piping the audio straight away (but adding it to an array of responses, which i then hope to concatenate together and create an audio file from).  

*within a `return new Promise()` block*  
```js

textToSpeech.synthesize(synthesizeParams, (err, audio) => {
    if (err) {
        console.log("an error occurred: " + err);
        return reject(err);
    }
    console.log(audio);
    resolve(audio);
});
```
or:
```js
textToSpeech
    .synthesize(synthesizeParams)
    .then(audio => {
        return audio;
    })
    .catch(err => {
        console.log("an error occurred: " + err);
    });
```

Should/can I be using `synthesizeStream.resume();` in this situation?

**Comment by dpopp07** [comment]
@oshihirii The `.resume()` function does not apply to your use case because you are using the HTTP version of Synthesize, not the WebSocket version. Only the WebSocket version returns a Stream.

If you want to use a Stream, you need to use the method `synthesizeUsingWebsocket`.

**Comment by oshihirii** [comment]
@dpopp07 - thank you.  I’m a bit confused about this as someone who seems quite knowledgeable about the matter has said in comments on this question that the response type is a stream:  https://stackoverflow.com/q/57157632.

**Comment by dpopp07** [comment]
I think I know where the confusion is coming from. We have two methods: `synthesize()`, which uses HTTP, and `synthesizeUsingWebSocket()`, which uses a WebSocket connection.

`synthesizeUsingWebSocket()` returns an actual stream.

`synthesize()` has a _response type_ of stream, but does not actually return a stream itself.

This means you need to handle the response, which is a stream. For example:

```js
textToSpeech.synthesize(synthesizeParams, (err, res) => {
  // res is a stream
  res.on('message', (message, data) => {
    resolve(data);
  });

  res.on('error', err => {
    console.log("an error occurred: " + err); 
    return reject(err);
  });
});
```
